### PR TITLE
Make dispatcher pluggable

### LINKED ISSUE #8834: Make Pulsar Dispatcher Pluggable
**Is your enhancement request related to a problem? Please describe.**
To support features like apache/pulsar#8544 without changing default dispatcher, we should to make the dispatcher pluggable. So user will be able to customize their own dispatcher can replace the default one.

**Describe the solution you'd like**
Add config to specify if customized dispatcher is needed for each subscription mode, as dispatcher is created per subscription mode. 
Add a dispatcher factory, if no customized dispatcher specified default one will be created, else customized one will be loaded with nar classloader.

### PR DESCRIPTION
**Pull request by MarvinCai**
Fixes #8834

### Motivation
Make dispatcher pluggable so we can support features like #8544 without changing default dispatcher.

### Modifications
Moved Dispatcher interface and abstract classes to dispatcher packages to make package hierarchy a bit cleaner. 
Add configuration for specifying customized dispatcher for persistent/nonpersistent + subtype combination.
Create util class to load customized dispatcher from NAR file.
Create factory class to supply dispatcher based on provided configuration.

### Verifying this change

Existing test cases passed, added new unit tests for util class and dispatcher factory.

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: no
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: no

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 1]
```diff
@@ -626,6 +626,62 @@
             + " non-backlog consumers as well.")
     private boolean dispatchThrottlingOnNonBacklogConsumerEnabled = false;
 
+    // <-- Pluggable dispatcher -->
```
Comment (zymap) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:629]: Do we need to add the following configurations into the `broker.conf` and `standalone.conf`?

**Review by zymap — COMMENTED** [review]
Does this need to update doc? Please open an issue to track doc if needed.

**Review by sijie — CHANGES_REQUESTED** [review]
@MarvinCai Thank you for putting this up!

But I think we should only have one DispatchFactory interface and it can be used for creating dispatchers for different subscription types. We shouldn't introduce that many configuration settings.

Also it might be worth writing a PIP for this effort.

**Review by rdhabalia — CHANGES_REQUESTED** [review]
@MarvinCai 

The goal of making dispatcher pluggable is to avoid any computation and complexity in the dispatch path. So, we should target to make dispatching-path pluggable rather than making dispatcher pluggable. I thinking making dispatcher pluggable will add a lot of work for the user and will be complex. It will also make implementation simpler without adding configuration for each dispatcher as well. So, we should create a PIP for this feature and discuss it.
We were also working on the same feature proposal. so, if it's fine then we can submit the proposal and start the discussion.

**Pull request closed without merging**